new-new-packages/compiler/src/language_server/references.rs

```diff
@@ -0,0 +1,183 @@
+        hir::{self, Body, Expression, HirDb, Lambda},
+    },
+    database::Database,
+    input::Input,
+};
+
+use super::utils::{LspPositionConversion, TupleToPosition};
+
+pub fn find_references(db: &Database, params: ReferenceParams) -> Option<Vec<Location>> {
+    let position = params.text_document_position;
+    let references = find(db, position.clone(), params.context.include_declaration)?
+        .into_iter()
+        .map(|it| Location {
+            uri: position.text_document.uri.clone(),
+            range: it.range,
+        })
+        .collect();
+    Some(references)
+}
+
+pub fn find_document_highlights(
+    db: &Database,
+    params: DocumentHighlightParams,
+) -> Option<Vec<DocumentHighlight>> {
+    find(db, params.text_document_position_params, true)
+}
+
+fn find(
+    db: &Database,
+    params: TextDocumentPositionParams,
+    include_declaration: bool,
+) -> Option<Vec<DocumentHighlight>> {
+    let input: Input = params.text_document.uri.clone().into();
+    let position = params.position;
+    let offset = db.offset_from_lsp(input.clone(), position.line, position.character);
+
+    let origin_cst = db.find_cst_by_offset(input.clone(), offset);
+    match origin_cst.kind {
+        CstKind::Identifier { .. } => {}
+        _ => return None,
+    }
+
+    let origin_hir_id = db.cst_to_hir_id(input.clone(), origin_cst.id)?;
+    Some(db.references(input, origin_hir_id, include_declaration))
+}
+
+#[salsa::query_group(ReferencesDbStorage)]
+pub trait ReferencesDb: HirDb + LspPositionConversion {
+    fn references(
+        &self,
+        input: Input,
+        id: hir::Id,
+        include_declaration: bool,
+    ) -> Vec<DocumentHighlight>;
+}
+
+fn references(
+    db: &dyn ReferencesDb,
+    input: Input,
+    id: hir::Id,
+    include_declaration: bool,
+) -> Vec<DocumentHighlight> {
+    let (hir, _) = db.hir(input.clone()).unwrap();
+
+    let mut context = Context::new(db, input, id, include_declaration);
+    context.visit_body(hir.as_ref());
+    context.references
+}
+
+struct Context<'a> {
+    db: &'a dyn ReferencesDb,
+    input: Input,
+    id: hir::Id,
+    include_declaration: bool,
+    discovered_references: HashSet<hir::Id>,
+    references: Vec<DocumentHighlight>,
+}
+impl<'a> Context<'a> {
+    fn new(db: &'a dyn ReferencesDb, input: Input, id: hir::Id, include_declaration: bool) -> Self {
+        Self {
+            db,
+            input,
+            id,
+            include_declaration,
+            discovered_references: HashSet::new(),
+            references: vec![],
+        }
+    }
+
+    fn visit_body(&mut self, body: &Body) {
+        if body.identifiers.contains_key(&self.id) {
+            self.add_reference(self.id.clone(), DocumentHighlightKind::WRITE);
+        }
+        for (id, expression) in &body.expressions {
+            self.visit_expression(id.to_owned(), expression);
+        }
+    }
+    fn visit_expressions(&mut self, ids: &[hir::Id]) {
```

Comment:
```suggestion
    fn visit_ids(&mut self, ids: &[hir::Id]) {
```

To be more consistent with the method below.